Add debug logging in ClassfileManager

gkossakowski opened the pull request:
Add logging of various operations the transactional class file manager is
doing. You can pass logger to be used by the transactional class file
manager by using overloaded definition of `ClassfileManager.transactional`
method. The old overload has been deprecated.

The factory methods for class file manager in IncOptions companion object
has been deprecated in favor of using ClassfileManager companion object
directly. The code in Defaults.scala has been updated to use non-deprecated
methods.

The canonical way of enabling transactional class file manager in sbt
project is:

``` scala
incOptions := incOptions.value.withNewClassfileManager(
  sbt.inc.ClassfileManager.transactional(
    crossTarget.value / "classes.bak",
    (streams in (Compile, compile)).value.log
  )
)
```

It's a bit verbose which shows that the api for this is not the best.
However, I don't expect sbt users to need this code very often.

This patch should help debug the problem described in #1184

gkossakowski commented:
Review by @jsuereth

jsuereth commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -215,7 +215,8 @@ object Defaults extends BuildCommon
 	def compileBase = inTask(console)(compilersSetting :: Nil) ++ compileBaseGlobal ++ Seq(
-		incOptions := IncOptions.setTransactional(incOptions.value, crossTarget.value / "classes.bak"),
+		incOptions := incOptions.value.withNewClassfileManager(
+			sbt.inc.ClassfileManager.transactional(crossTarget.value / "classes.bak", sbt.Logger.Null)),
```
should this one be `streams.value.log` instead of `sbt.Logger.Null` ?

gkossakowski replied:
Logging is opt-in. I didn't want to clutter the debug log too much. Most of the time you don't really need to know what ClassfileManager does even if you are debugging incremental compiler problems.

jsuereth commented:
your example may want to show:

``` scala
(streams in (compile, Compile)).value.log
```

Or some such variant if you want the log to align with the `compile in Compile` task.   The stream-logging stuff is a bit... fun right now, and needs some cleanup.  But just curious if the output locations is:

a. where you want it
b. in the same file as the `compile in Compile` logs.

jsuereth commented:
Oh, otherwise LGTM

gkossakowski committed:
Add debug logging in ClassfileManager

Add logging of various operations the transactional class file manager is
doing. You can pass logger to be used by the transactional class file
manager by using overloaded definition of `ClassfileManager.transactional`
method. The old overload has been deprecated.

The factory methods for class file manager in IncOptions companion object
has been deprecated in favor of using ClassfileManager companion object
directly. The code in Defaults.scala has been updated to use non-deprecated
methods. The logging is turned off by default.

The canonical way of enabling transactional class file manager in sbt
project is:

```
incOptions := incOptions.value.withNewClassfileManager(
  sbt.inc.ClassfileManager.transactional(
    crossTarget.value / "classes.bak",
    (streams in (compile, Compile)).value.log
  )
)
```

It's a bit verbose which shows that the api for this is not the best.
However, I don't expect sbt users to need this code very often.

This patch should help debug the problem described in #1184

gkossakowski commented:
Clarified in commit message that logging is off by default and fixed the code snippet according to your suggestion.

I agree that logging in sbt could receive some love.

aloiscochard commented:
@gkossakowski the instruction you provided did not work.

I had to do:

```
    incOptions := incOptions.value.withNewClassfileManager(
      sbt.inc.ClassfileManager.transactional(
        crossTarget.value / "classes.bak",
        (streams in Compile).value.log
      )
    )
```

I prefer not to know why it should be like that, but I thought it was worth mentioning if anyone else try to use this.

gkossakowski commented:
@jsuereth: any ideas on why the original snippet would not work?

jsuereth commented:
@gkossakowski I sent you an email about it elsewhere.   You have the wrong order for `in` so you confuse the hell out of scala looking for implicits on the overloaded method.

Your code _should_ have been:

```
incOptions := incOptions.value.withNewClassfileManager(
  sbt.inc.ClassfileManager.transactional(
    crossTarget.value / "classes.bak",
    (streams in (Compile, compile)).value.log
  )
)
```

See me note above.

gkossakowski commented:
@jsuereth: I updated the description in this issue. Thanks.
